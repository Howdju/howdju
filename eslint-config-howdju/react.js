module.exports = {
  parserOptions: {
    "ecmaFeatures": {
      "jsx": true
    },
  },
  env: {
    "browser": true,
  },
  extends: [
    "plugin:react/recommended"
  ],
  plugins: [
    "react",
  ],
  rules: {
    indent: [
      "warn",
      2,
      {
        "SwitchCase": 1,
        "ignoredNodes": [
          // We indent to the indent of the first prop, which this rule doesn't support
          "JSXAttribute",
          "JSXSpreadAttribute",
        ],
      }
    ],
    "react/boolean-prop-naming": "error",
    "react/no-children-prop": [
      // react-md requires us to use a children prop
      "off",
    ],
    "react/prop-types": [
      // This rule can't pick up props added by redux's mapStateToProps/mapDispatchToProps
      "off",
    ],
    // This doesn't seem to help; need to disable indent on JSXAttribute
    // "react/jsx-indent-props": [
    //   // We indent to the indent of the first prop, which this rule doesn't support
    //   "off"
    // ],
    "require-yield": [
      // Some of our sagas are side-effect only, requiring no yields.  And I don't think there's any harm in this.
      "off"
    ],
  },
  settings: {
    react: {
      version: "detect",
    },
  },
};
