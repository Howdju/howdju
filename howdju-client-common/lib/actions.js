import { createAction as actionCreator } from 'redux-actions'

/** react-actions has a convention that its action creators .toString return the action type.
 * .toString appears to happen automatically when an action creator is the key of an object, but in some
 * cases we will need to call .toString manually.  This method will help us locate those places in the code
 */
export const str = ac => ac.toString()

export const extension = {
  focusJustificationOnUrl: actionCreator('EXTENSION/FOCUS_JUSTIFICATION_ON_URL',
    (url, justificationId, howdjuUrl) => ({url, justificationId, howdjuUrl}))
}

export const extensionFrame = {
  createJustification: actionCreator('EXTENSION_FRAME/CREATE_JUSTIFICATION')
}
