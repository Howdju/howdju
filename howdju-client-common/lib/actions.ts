import { ActionCreator } from "redux";
import { createAction } from "redux-actions";
import {
  CounteredJustification,
  decircularizeJustification,
  Justification,
  Url,
  WritQuote,
} from "howdju-common";

/**
 * redux-actions and @reduxjs/toolkit have a convention that action creators get a .toString method
 * that returns the action type. The call to .toString appears to happen automatically when an
 * action creator is the key of an object, but:
 *
 * 1) TypeScript doesn't recognize action creators as valid object keys, and
 * 2) The reducers/sagas keyed in this way cannot receive inferred types.
 *
 * We can use this method to satisfy (1). Because of (2), though, we will probably eventually want
 * to phase out this function.
 */
export const str = (ac: ActionCreator<unknown>) => ac.toString();

// Actions to send to the extension
export const extension = {
  highlightTarget: createAction(
    "EXTENSION/HIGHLIGHT_TARGET",
    (
      justification: CounteredJustification,
      writQuote: WritQuote,
      url: Url
    ) => ({
      justification: decircularizeJustification(justification),
      writQuote,
      url,
    })
  ),
  messageHandlerReady: createAction("EXTENSION/MESSAGE_HANDLER_READY"),
};

// Actions to send to the frame the extension inserts that is hosting the web app
export const extensionFrame = {
  createJustification: createAction("EXTENSION_FRAME/CREATE_JUSTIFICATION"),
  gotoJustification: createAction(
    "EXTENSION_FRAME/GOTO_JUSTIFICATION",
    (justification: Justification) => ({ justification })
  ),
  ackMessage: createAction("EXTENSION_FRAME/ACK_MESSAGE"),
};
