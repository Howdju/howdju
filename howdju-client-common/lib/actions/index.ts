export * from "./actionHelpers";
export * from "./apiActions";
export * from "./apiActionHelpers";
export * from "./apiActionTypes";
