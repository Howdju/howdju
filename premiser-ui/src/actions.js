import { createAction as actionCreator } from 'redux-actions'
import {
  JustificationVotePolarity,
  decircularizeJustification,
  makeStatementTagVote,
  StatementTagVotePolarity,
} from "howdju-common"
import reduce from 'lodash/reduce'
import mapValues from 'lodash/mapValues'
import assign from 'lodash/assign'

const actionTypeDelim = '/'

/** react-actions has a convention that its action creators .toString return the action type.
 * .toString appears to happen automatically when an action creator is the key of an object, but in some
 * cases we will need to call .toString manually.  This method will help us locate those places in the code
 */
export const str = ac => ac.toString()

/** Helper to easily allow us to pass an object with 'action groups' to redux-react's connect method.
 * Action groups are what we call the objects below with react-actions action creators.  They are just
 * a way to organize related action creators.  The convention with redux-react's connect method's mapDispatchToProps
 * is to pass an object with keys equal to action creators.  redux-react will automatically turn the action creators
 * into dispatched versions.  This helper accomplishes the same for an object the properties of which are action creator
 * groups like those defined below.
 */
export const mapActionCreatorGroupToDispatchToProps = (actionCreatorGroups, otherActions) => (dispatch) => {
  const dispatchProps = mapValues(actionCreatorGroups, (actionCreatorGroup) =>
    mapValues(actionCreatorGroup, (actionCreator) =>
      (...args) => dispatch(actionCreator.apply(null, args))
    )
  )

  if (otherActions) {
    assign(dispatchProps, mapValues(otherActions, actionCreator => (...args) => dispatch(actionCreator.apply(null, args))))
  }

  return dispatchProps
}

/** Create an action creator having a property `.response` with another action creator for corresponding API responses */
const apiActionCreator = (...args) => {
  const actionType = 'API' + actionTypeDelim + args[0]
  const ac = actionCreator(actionType, ...args.slice(1))
  ac.response = actionCreator(actionType + actionTypeDelim + 'RESPONSE', (payload, meta) => payload, (payload, meta) => meta)
  return ac
}

export const app = {
  searchMainSearch: actionCreator('APP/SEARCH_MAIN_SEARCH', searchText => ({searchText})),
}

export const api = {
  callApi: apiActionCreator('CALL_API'),
  fetchStatement: apiActionCreator('FETCH_STATEMENT', (statementId) => ({statementId})),
  fetchStatements: apiActionCreator('FETCH_STATEMENTS', (statementIds) => ({statementIds})),
  fetchStatementCompound: apiActionCreator('FETCH_STATEMENT_COMPOUND', (statementCompoundId) => ({statementCompoundId})),
  fetchStatementJustifications: apiActionCreator('FETCH_STATEMENT_JUSTIFICATIONS', (statementId) => ({statementId})),
  fetchWritQuote: apiActionCreator('FETCH_WRIT_QUOTE', (writQuoteId) => ({writQuoteId})),
  fetchJustificationBasisCompound: apiActionCreator('FETCH_JUSTIFICATION_BASIS_COMPOUND',
    (justificationBasisCompoundId) => ({justificationBasisCompoundId})),
  fetchSourceExcerptParaphrase: apiActionCreator('FETCH_SOURCE_EXCERPT_PARAPHRASE',
    (sourceExcerptParaphraseId) => ({sourceExcerptParaphraseId})),

  fetchRecentStatements: apiActionCreator('FETCH_RECENT_STATEMENTS', (widgetId, count, continuationToken) => ({widgetId, count, continuationToken})),
  fetchRecentWrits: apiActionCreator('FETCH_RECENT_WRITS', (widgetId, count, continuationToken) => ({widgetId, continuationToken, count})),
  fetchRecentWritQuotes: apiActionCreator('FETCH_RECENT_WRIT_QUOTES', (widgetId, count, continuationToken) => ({widgetId, count, continuationToken})),
  fetchRecentJustifications: apiActionCreator('FETCH_RECENT_JUSTIFICATIONS', (widgetId, count, continuationToken) => ({widgetId, count, continuationToken})),

  fetchFeaturedPerspectives: apiActionCreator('FETCH_FEATURED_PERSPECTIVES', (widgetId) => ({widgetId})),

  fetchJustificationsSearch: apiActionCreator('FETCH_JUSTIFICATIONS_SEARCH',
    ({filters, count, continuationToken}) => ({filters, count, continuationToken})
  ),
  login: apiActionCreator('LOGIN', (credentials) => ({credentials})),
  logout: apiActionCreator('LOGOUT'),

  verifyJustification: apiActionCreator('VERIFY_JUSTIFICATION', (justification) => ({
    justificationVote: {
      justificationId: justification.id,
      polarity: JustificationVotePolarity.POSITIVE,
    },
    previousJustificationVote: justification.vote,
  })),
  unVerifyJustification: apiActionCreator('UN_VERIFY_JUSTIFICATION', (justification) => ({
    justificationVote: {
      justificationId: justification.id,
      polarity: JustificationVotePolarity.POSITIVE,
    },
    previousJustificationVote: justification.vote,
  })),
  disverifyJustification: apiActionCreator('DISVERIFY_JUSTIFICATION', (justification) => ({
    justificationVote: {
      justificationId: justification.id,
      polarity: JustificationVotePolarity.NEGATIVE,
    },
    previousJustificationVote: justification.vote,
  })),
  unDisverifyJustification: apiActionCreator('UN_DISVERIFY_JUSTIFICATION', (justification) => ({
    justificationVote: {
      justificationId: justification.id,
      polarity: JustificationVotePolarity.NEGATIVE,
    },
    previousJustificationVote: justification.vote,
  })),

  tagStatement: apiActionCreator('TAG_STATEMENT', (statementId, tag, statementTagVote) => ({
    statementTagVote: makeStatementTagVote({
      polarity: StatementTagVotePolarity.POSITIVE,
      statement: {id: statementId},
      tag,
    }),
    prevStatementTagVote: statementTagVote,
  })),
  antiTagStatement: apiActionCreator('ANTI_TAG_STATEMENT', (statementId, tag, statementTagVote) => ({
    statementTagVote: makeStatementTagVote({
      polarity: StatementTagVotePolarity.NEGATIVE,
      statement: {id: statementId},
      tag,
    }),
    prevStatementTagVote: statementTagVote,
  })),
  unTagStatement: apiActionCreator('UN_TAG_STATEMENT', (statementTagVote) => ({prevStatementTagVote: statementTagVote})),

  createStatement: apiActionCreator('CREATE_STATEMENT', (statement) => ({statement})),
  updateStatement: apiActionCreator('UPDATE_STATEMENT', (statement) => ({statement}), (s, nonce) => ({nonce})),
  deleteStatement: apiActionCreator('DELETE_STATEMENT', (statement) => ({statement})),

  fetchStatementTextSuggestions: apiActionCreator('FETCH_STATEMENT_TEXT_SUGGESTIONS', (statementText, suggestionsKey) => ({
    statementText,
    suggestionsKey,
  })),
  cancelStatementTextSuggestions: apiActionCreator('CANCEL_STATEMENT_TEXT_SUGGESTIONS', (suggestionsKey) => ({
    cancelTarget: str(api.fetchStatementTextSuggestions),
    suggestionsKey,
  })),

  fetchWritTitleSuggestions: apiActionCreator('FETCH_WRIT_TITLE_SUGGESTIONS', (writTitle, suggestionsKey) => ({
    writTitle,
    suggestionsKey,
  })),
  cancelWritTitleSuggestions: apiActionCreator('CANCEL_WRIT_TITLE_SUGGESTIONS', (suggestionsKey) => ({
    cancelTarget: str(api.fetchWritTitleSuggestions),
    suggestionsKey,
  })),

  fetchTagNameSuggestions: apiActionCreator('FETCH_TAG_NAME_SUGGESTIONS', (tagName, suggestionsKey) => ({
    tagName,
    suggestionsKey,
  })),
  cancelTagNameSuggestions: apiActionCreator('CANCEL_TAG_NAME_SUGGESTIONS', (suggestionsKey) => ({
    cancelTarget: str(api.fetchTagNameSuggestions),
    suggestionsKey,
  })),

  fetchMainSearchSuggestions: apiActionCreator('FETCH_MAIN_SEARCH_SUGGESTIONS', (searchText, suggestionsKey) => ({
    searchText,
    suggestionsKey
  })),
  cancelMainSearchSuggestions: apiActionCreator('CANCEL_MAIN_SEARCH_SUGGESTIONS', (suggestionsKey) => ({
    cancelTarget: str(api.fetchMainSearchSuggestions),
    suggestionsKey,
  })),

  createJustification: apiActionCreator('CREATE_JUSTIFICATION', (justification) => ({justification: decircularizeJustification(justification)})),
  updateWritQuote: apiActionCreator('UPDATE_WRIT_QUOTE', (writQuote) => ({writQuote})),
  deleteJustification: apiActionCreator('DELETE_JUSTIFICATION', (justification) => ({justification})),

  fetchMainSearchResults: apiActionCreator('FETCH_MAIN_SEARCH_RESULTS', (searchText) => ({searchText})),
  fetchStatementsSearch: apiActionCreator('FETCH_STATEMENTS_SEARCH', (searchText) => ({searchText})),

  fetchTag: apiActionCreator('FETCH_TAG', (tagId) => ({tagId})),
  fetchTaggedStatements: apiActionCreator('FETCH_TAGGED_STATEMENTS', (tagId) => ({tagId})),
}
export const apiActionCreatorsByActionType = reduce(api, (result, actionCreator) => {
  result[actionCreator] = actionCreator
  return result
}, {})

export const ui = {
  unhandledAppClick: actionCreator('UI/UNHANDLED_APP_CLICK'),
  unhandledAppTouch: actionCreator('UI/UNHANDLED_APP_TOUCH'),
  showNavDrawer: actionCreator('SHOW_NAV_DRAWER'),
  hideNavDrawer: actionCreator('HIDE_NAV_DRAWER'),
  toggleNavDrawerVisibility: actionCreator('TOGGLE_NAV_DRAWER_VISIBILITY'),
  setNavDrawerVisibility: actionCreator('SET_NAV_DRAWER_VISIBILITY'),
  addToast: actionCreator('ADD_TOAST', (text) => ({text})),
  dismissToast: actionCreator('DISMISS_TOAST'),

  showNewJustificationDialog: actionCreator('SHOW_NEW_JUSTIFICATION_DIALOG'),
  hideNewJustificationDialog: actionCreator('HIDE_NEW_JUSTIFICATION_DIALOG'),

  mainSearchTextChange: actionCreator('MAIN_SEARCH_TEXT_CHANGE'),
  loginCredentialChange: actionCreator('LOGIN_CREDENTIAL_CHANGE'),
  clearJustificationsSearch: actionCreator('UI/CLEAR_JUSTIFICATIONS_SEARCH'),

  clearRecentStatements: actionCreator('UI/CLEAR_RECENT_STATEMENTS', (widgetId) => ({widgetId})),
  clearRecentWrits: actionCreator('UI/CLEAR_RECENT_WRITS', (widgetId) => ({widgetId})),
  clearRecentWritQuotes: actionCreator('UI/CLEAR_RECENT_WRIT_QUOTES', (widgetId) => ({widgetId})),
  clearRecentJustifications: actionCreator('UI/CLEAR_RECENT_JUSTIFICATIONS', (widgetId) => ({widgetId})),

  beginInteractionWithTransient: actionCreator('UI/BEGIN_INTERACTION_WITH_TRANSIENT', (transientId) => ({transientId})),
  endInteractionWithTransient: actionCreator('UI/END_INTERACTION_WITH_TRANSIENT', (transientId) => ({transientId})),
  showTransient: actionCreator('UI/SHOW_TRANSIENT', (transientId) => ({transientId})),
  scheduleDelayedHideTransient: actionCreator('UI/SCHEDULE_DELAYED_HIDE_TRANSIENT', (transientId, hideDelay) => ({transientId, hideDelay})),
  tryCancelDelayedHideTransient: actionCreator('UI/TRY_CANCEL_DELAYED_HIDE_TRANSIENT',
    (transientId, cause) => ({transientId}),
    (transientId, cause) => ({cause}),
  ),
  cancelDelayedHideTransient: actionCreator('UI/CANCEL_DELAYED_HIDE_TRANSIENT', (transientId) => ({transientId})),
  hideAllTransients: actionCreator('UI/HIDE_ALL_TRANSIENTS'),
  hideOtherTransients: actionCreator('UI/HIDE_OTHER_TRANSIENTS', (visibleTransientId) => ({visibleTransientId})),
  hideTransient: actionCreator('UI/HIDE_TRANSIENT',
    (transientId, cause) => ({transientId}),
    (transientId, cause) => ({cause}),
  ),
  windowResize: actionCreator('UI/WINDOW_RESIZE'),

  expand: actionCreator('UI/EXPAND', (widgetId) => ({widgetId})),
  collapse: actionCreator('UI/COLLAPSE', (widgetId) => ({widgetId})),

  enableMobileSite: actionCreator('UI/ENABLE_MOBILE_SITE'),
  disableMobileSite: actionCreator('UI/DISABLE_MOBILE_SITE'),

  clearTaggedStatements: actionCreator('UI/CLEAR_TAGGED_STATEMENTS'),
}

const commitEdit = actionCreator('EDITORS/COMMIT_EDIT', (editorType, editorId) => ({editorType, editorId}))
commitEdit.result = actionCreator('EDITORS/COMMIT_EDIT' + actionTypeDelim + 'RESULT',
  (editorType, editorId, result) => ({editorType, editorId, result}),
  (...args) => {
    if (args.length === 4) {
      return args[3]
    } else if (args.length === 2 && args[0] instanceof Error) {
      return args[1]
    }
    return undefined
  }
)
export const editors = {
  beginEdit: actionCreator('EDITORS/BEGIN_EDIT', (editorType, editorId, entity) => ({editorType, editorId, entity})),
  propertyChange: actionCreator('EDITORS/PROPERTY_CHANGE', (editorType, editorId, properties) => ({editorType, editorId, properties})),
  commitEdit,
  cancelEdit: actionCreator('EDITORS/CANCEL_EDIT', (editorType, editorId) => ({editorType, editorId})),

  addUrl: actionCreator('EDITORS/ADD_URL', (editorType, editorId) => ({editorType, editorId})),
  removeUrl: actionCreator('EDITORS/REMOVE_URL', (editorType, editorId, url, index) => ({
    editorType,
    editorId,
    url,
    index
  })),

  addStatementCompoundAtom: actionCreator('EDITORS/ADD_STATEMENT_COMPOUND_ATOM', (editorType, editorId, index) => ({
    editorType,
    editorId,
    index,
  })),
  removeStatementCompoundAtom: actionCreator('EDITORS/REMOVE_STATEMENT_COMPOUND_ATOM', (editorType, editorId, atom, index) => ({
    editorType,
    editorId,
    atom,
    index
  })),

  addJustificationBasisCompoundAtom: actionCreator('EDITORS/ADD_JUSTIFICATION_BASIS_COMPOUND_ATOM',
    (editorType, editorId, index) => ({
      editorType,
      editorId,
      index,
    })
  ),
  removeJustificationBasisCompoundAtom: actionCreator('EDITORS/REMOVE_JUSTIFICATION_BASIS_COMPOUND_ATOM',
    (editorType, editorId, atom, index) => ({
      editorType,
      editorId,
      atom,
      index
    })
  ),
  addJustificationBasisCompoundAtomSourceExcerptParaphraseWritQuoteUrl: actionCreator(
    'EDITORS/ADD_JUSTIFICATION_BASIS_COMPOUND_ATOM_SOURCE_EXCERPT_PARAPHRASE_WRIT_QUOTE_URL',
    (editorType, editorId, atomIndex, urlIndex) => ({editorType, editorId, atomIndex, urlIndex})
  ),
  removeJustificationBasisCompoundAtomSourceExcerptParaphraseWritQuoteUrl: actionCreator(
    'EDITORS/REMOVE_JUSTIFICATION_BASIS_COMPOUND_ATOM_SOURCE_EXCERPT_PARAPHRASE_WRIT_QUOTE_URL',
    (editorType, editorId, atom, atomIndex, url, urlIndex) => ({editorType, editorId, atom, atomIndex, url, urlIndex})
  ),

  tagStatement: actionCreator('EDITORS/TAG_STATEMENT', (editorType, editorId, tag) => ({editorType, editorId, tag})),
  unTagStatement: actionCreator('EDITORS/UN_TAG_STATEMENT', (editorType, editorId, tag) => ({editorType, editorId, tag})),
  antiTagStatement: actionCreator('EDITORS/ANTI_TAG_STATEMENT', (editorType, editorId, tag) => ({editorType, editorId, tag})),
}

export const goto = {
  login: actionCreator('GOTO/LOGIN', (loginRedirectLocation) => ({loginRedirectLocation})),
  statement: actionCreator('GOTO/STATEMENT', (statement) => ({statement})),
  mainSearch: actionCreator('GOTO/MAIN_SEARCH', (mainSearchText) => ({mainSearchText})),
  tag: actionCreator('GOTO/TAG', (tag) => ({tag})),
}

export const flows = {
  fetchAndBeginEditOfNewJustificationFromBasisSource: actionCreator(
    'FLOWS/FETCH_AND_BEGIN_EDIT_OF_NEW_JUSTIFICATION_FROM_BASIS_SOURCE',
    (editorType, editorId, basisSourceType, basisSourceId) => ({editorType, editorId, basisSourceType, basisSourceId})
  ),
  commitEditThenView: actionCreator('FLOWS/COMMIT_STATEMENT_THEN_VIEW',
    (editorType, editorId) => ({editorType, editorId})),
  commitEditThenPutActionOnSuccess: actionCreator('FLOWS/COMMIT_EDIT_THEN_PUT_ACTION_ON_SUCCESS',
    (editorType, editorId, onSuccessAction) => ({editorType, editorId, onSuccessAction})),
}

export const autocompletes = {
  clearSuggestions: actionCreator('AUTOCOMPLETES/CLEAR_SUGGESTIONS', (suggestionsKey) => ({suggestionsKey}))
}

export const errors = {
  clearLoggedErrors: actionCreator('ERRORS/CLEAR_LOGGED_ERRORS'),
}