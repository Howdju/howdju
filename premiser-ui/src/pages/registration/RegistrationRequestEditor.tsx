import { CreateRegistrationRequestInput } from "howdju-common";

import withEditor from "@/editors/withEditor";
import { EditorTypes } from "@/reducers/editors";
import RegistrationRequestFields from "./RegistrationRequestEditorFields";
import { CreateRegistrationRequestConfig } from "@/sagas/editors/editorCommitEditSaga";

/** A RegistrationConfirmation editor. */
export default withEditor(
  EditorTypes.REGISTRATION_REQUEST,
  RegistrationRequestFields,
  "registrationRequest",
  CreateRegistrationRequestInput,
  CreateRegistrationRequestConfig
);
