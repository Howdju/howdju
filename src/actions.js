export const CREATE_STATEMENT = 'CREATE_STATEMENT'
export const JUSTIFY_STATEMENT = 'JUSTIFY_STATEMENT'
export const CREATE_JUSTIFIED_STATEMENT = 'CREATE_JUSTIFIED_STATEMENT'
export const ACCEPT_JUSTIFICATION = 'ACCEPT_JUSTIFICATION'
export const REJECT_JUSTIFICATION = 'REJECT_JUSTIFICATION'

export function createStatement(text) {
  return { type: CREATE_STATEMENT, payload: { text }}
}

export function justifyStatement(statementId, justification) {
  return { type: JUSTIFY_STATEMENT, payload: { statementId, justification }}
}

export function createJustifiedStatement(statement, justifications) {
  return { type: CREATE_JUSTIFIED_STATEMENT, payload: { statement, justifications }}
}

export function acceptJustification(justificationId) {
  return { type: ACCEPT_JUSTIFICATION, payload: { justificationId }}
}

export function rejectJustification(justificationId) {
  return { type: REJECT_JUSTIFICATION, payload: { justificationId }}
}