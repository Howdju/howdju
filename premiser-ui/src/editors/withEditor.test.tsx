import React from "react";
import { screen } from "@testing-library/react";

import {
  DISABLED_BUTTON_CLASS,
  renderWithProviders,
  setupDefaultStore,
  setupUserEvent,
  withFakeTimers,
} from "@/testUtils";
import { editors } from "@/actions";
import RegistrationRequestEditor from "@/pages/registration/RegistrationRequestEditor";

const REACT_MD_LABEL_ERROR_CLASS = "rmd-label--error";

withFakeTimers();

describe("withEditor", () => {
  describe("Exemplary component (RegistrationRequestEditor)", () => {
    test("submit button appears disabled when edit entity is invalid", () => {
      const editorId = "test-editor-id";

      const { preloadedState, history, store } = setupDefaultStore();
      store.dispatch(
        editors.beginEdit("REGISTRATION_REQUEST", editorId, { email: "" })
      );

      // Act
      renderWithProviders(
        <RegistrationRequestEditor
          id="test-writ-quote-editor"
          editorId={editorId}
        />,
        { preloadedState, history, store }
      );

      // Assert
      const submitButton = screen.getByRole("button", { name: /save/i });
      expect(submitButton).toHaveClass(DISABLED_BUTTON_CLASS);
    });

    test("submitting invalid entity shows errors", async () => {
      const user = setupUserEvent();
      const editorId = "test-editor-id";

      const { preloadedState, history, store } = setupDefaultStore();
      store.dispatch(
        editors.beginEdit("REGISTRATION_REQUEST", editorId, { email: "" })
      );
      const { container } = renderWithProviders(
        <RegistrationRequestEditor
          id="test-writ-quote-editor"
          editorId={editorId}
        />,
        { preloadedState, history, store }
      );

      // Act
      await user.click(screen.getByRole("button", { name: /save/i }));

      // Assert
      const titleInput = screen.getByLabelText(/email/i);
      const titleLabel = container.querySelector(`[for=${titleInput.id}]`);
      expect(titleLabel).toHaveClass(REACT_MD_LABEL_ERROR_CLASS);

      expect(screen.getByText(/invalid email/i)).toBeInTheDocument();

      jest.runAllTimers();
      expect(container).toMatchSnapshot();
    });
  });
});
