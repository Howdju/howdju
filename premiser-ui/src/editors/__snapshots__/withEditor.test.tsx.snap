// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withEditor Exemplary component (RegistrationRequestEditor) submitting invalid entity shows errors 1`] = `
<div>
  <form>
    <div
      class="rmd-card__content rmd-card__content--padded rmd-card__content--extra-padding rmd-card__content--remove-margin rmd-card__content--secondary"
    >
      <ul
        class="error-message"
      />
      <div
        class="md-text-field-container md-full-width md-text-field-container--multiline"
      >
        <label
          class="md-floating-label md-floating-label--inactive md-floating-label--inactive-sized md-text--error"
          for="email"
        >
          Email *
        </label>
        <div
          class="md-text-field-multiline-container md-text-field--floating-margin"
          style="height: 0px;"
        >
          <span
            aria-hidden="true"
          />
          <textarea
            aria-hidden="true"
            class="md-text-field md-text-field--multiline md-full-width md-text md-text-field--multiline-mask"
            readonly=""
            rows="1"
            tabindex="-1"
          />
          <textarea
            autocomplete="email"
            class="md-text-field md-text-field--multiline md-full-width md-text"
            id="email"
            name="email"
            required=""
            rows="1"
            style="height: 0px;"
          />
        </div>
        <hr
          class="md-divider md-divider--text-field md-divider--expand-from-left md-divider--text-field-error"
        />
        <div
          class="md-text-field-message-container md-full-width md-text--error"
        >
          <div
            aria-hidden="false"
            class="md-text-field-message md-text-field-message--active"
          >
            Invalid email
          </div>
          <div
            aria-hidden="true"
            class="md-text-field-message md-text-field-message--inactive md-text-field-message--counter"
          >
            0 / 512
          </div>
        </div>
      </div>
    </div>
    <div
      class="rmd-card__actions"
    >
      <button
        class="rmd-button rmd-button--text rmd-button--text-primary"
        type="button"
      >
        Cancel
        <span
          class="rmd-ripple-container"
        />
      </button>
      <button
        aria-disabled="true"
        class="rmd-button rmd-button--text rmd-button--contained rmd-button--disabled"
        title="Please correct the errors to continue"
        type="submit"
      >
        Save
        <span
          class="rmd-ripple-container"
        />
      </button>
    </div>
  </form>
</div>
`;
